### PR TITLE
[FLINK-8441] [State Backend] [RocksDB] change RocksDBListState to serialize values and separators in stream to avoid extra bytes copying

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

Currently, `RocksDBListState#update()` and `addAll()` will both serialize values into a list of bytes, manually merge the list of bytes with separators to an array of bytes, and then write to RocksDB. It results in extra step of copying bytes back and forth.

## Brief change log

Changed `RocksDBListState#update()` and `addAll()` to serialize values and separators directly into `keySerializationStream`, which avoids extra copying.

## Verifying this change

This change is already covered by existing tests, such as `RocksDBStateBackendTest`

## Does this pull request potentially affect one of the following parts:

none

## Documentation

none

cc @StefanRRichter

### CODE REVIEW

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBListState.java [hunk 1]
```diff
@@ -60,6 +59,11 @@
 	 */
 	private final WriteOptions writeOptions;
 
+	/**
+	 * Simulated separator of StringAppendTestOperator in RocksDB.
```
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBListState.java]: I would drop the word `Simulated`.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBListState.java [hunk 2]
```diff
@@ -202,13 +206,17 @@ public void addAll(List<V> values) throws Exception {
 	private byte[] getPreMergedValue(List<V> values) throws IOException {
 		DataOutputViewStreamWrapper out = new DataOutputViewStreamWrapper(keySerializationStream);
 
-		List<byte[]> bytes = new ArrayList<>(values.size());
+		keySerializationStream.reset();
+		byte first = 0;
```
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBListState.java]: I think `boolean` is a more appropriate type.

**Comment by bowenli86** [comment]
Make sense. Updated the PR also squashed all commits

**Comment by StefanRRichter** [comment]
Thanks for your contribution, I will merge this with some small cleanup.

**Pull request closed without merging**